miniapp/miniapp_cholesky.cpp

```diff
@@ -0,0 +1,241 @@
+
+T analytical_input_matrix(const GlobalElementIndex& index);
+T analytical_result_matrix(const GlobalElementIndex& index);
+
+void setup_input_matrix(Matrix<T, Device::CPU>& matrix) {
+  using namespace dlaf_test::matrix_test;
+
+  util_matrix::assertSizeSquare(matrix, __PRETTY_FUNCTION__, "matrix");
+  util_matrix::assertBlocksizeSquare(matrix, __PRETTY_FUNCTION__, "matrix");
+
+  set(matrix, analytical_input_matrix);
+}
+
+void cholesky_check(Matrix<T, Device::CPU>& matrix) {
+  using namespace dlaf_test;
+
+  CHECK_MATRIX_NEAR(analytical_result_matrix, matrix,
+                    4 * (matrix.size().rows() + 1) * TypeUtilities<T>::error,
+                    4 * (matrix.size().rows() + 1) * TypeUtilities<T>::error);
+}
+
+int hpx_main(hpx::program_options::variables_map& vm) {
+  options_t opts = check_options(vm);
+
+  comm::Communicator world(MPI_COMM_WORLD);
+  comm::CommunicatorGrid comm_grid(world, opts.grid_rows, opts.grid_cols, common::Ordering::ColumnMajor);
+
+  // Allocate memory for the matrix
+  GlobalElementSize matrix_size(opts.m, opts.m);
+  TileElementSize block_size(opts.mb, opts.mb);
+
+  Matrix<T, Device::CPU> matrix(matrix_size, block_size, comm_grid);
+  auto distribution = matrix.distribution();
+
+  for (auto run_index = 0; run_index < opts.nruns; ++run_index) {
+    if (0 == world.rank())
+      std::cout << "[" << run_index << "]" << std::endl;
+
+    setup_input_matrix(matrix);
+
+    // wait all setup tasks before starting benchmark
+    {
+      for (int local_tile_j = 0; local_tile_j < distribution.localNrTiles().cols(); ++local_tile_j)
+        for (int local_tile_i = 0; local_tile_i < distribution.localNrTiles().rows(); ++local_tile_i)
+          matrix(LocalTileIndex{local_tile_i, local_tile_j}).get();
+      MPI_Barrier(world);
+    }
+
+    common::Timer<> timeit;
+    cholesky(comm_grid, blas::Uplo::Lower, matrix);
+
+    // wait for last task and barrier for all ranks
+    {
+      GlobalTileIndex last_tile(matrix.nrTiles().rows() - 1, matrix.nrTiles().cols() - 1);
+      if (matrix.rankIndex() == distribution.rankGlobalTile(last_tile))
+        matrix(last_tile).get();
+
+      MPI_Barrier(world);
+    }
+    auto elapsed_time = timeit.elapsed();
+
+    double gigaflops;
+    {
+      double n = matrix.size().rows();
+      double add_mul = n * n * n / 6;
+      static_assert(!std::is_same<T, double>::value || !std::is_same<T, float>::value, "FIX GIGAFLOPS");
+      gigaflops = 2 * add_mul / elapsed_time / 1e9;
+    }
+
+    // print benchmark results
+    if (0 == world.rank())
+      std::cout << "[" << run_index << "]"
+                << " " << elapsed_time << "s"
+                << " " << gigaflops << "GFlop/s"
+                << " " << matrix.size() << " " << matrix.blockSize() << " " << comm_grid.size() << " "
+                << hpx::get_os_thread_count() << std::endl;
+
+    // (optional) run test
+    if (opts.do_check)
+      cholesky_check(matrix);
```

Comment:
same as for `set`...

Reply:
As per discussion we had, the current version is very easily optimizable, but we need to decide where to put it:
- in `dlaf_test`?
- here in the mini-app?
- or do we want to add an optimized library also for checks as we are doing for set methods in #135?

Reply:
As per discussion, we decided to not implement any optimized version of this check, for now it is not needed.